pkg/apis/machine/types.go

```diff
@@ -1117,3 +1117,75 @@ type GCPServiceAccount struct {
 	Email  string
 	Scopes []string
 }
+
+/********************** AlicloudMachineClass APIs ***************/
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlicloudMachineClass
+type AlicloudMachineClass struct {
+	// +optional
+	metav1.ObjectMeta
+
+	// +optional
+	metav1.TypeMeta
+
+	// +optional
+	Spec AlicloudMachineClassSpec
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AlicloudMachineClassList is a collection of AlicloudMachineClasses.
+type AlicloudMachineClassList struct {
+	// +optional
+	metav1.TypeMeta
+
+	// +optional
+	metav1.ListMeta
+
+	// +optional
+	Items []AlicloudMachineClass
+}
+
+// AlicloudMachineClassSpec is the specification of a cluster.
+type AlicloudMachineClassSpec struct {
+	ImageId                 string
+	InstanceType            string
+	Region                  string
+	ZoneId                  string
+	SecurityGroupId         string
+	VSwitchId               string
+	PrivateIpAddress        string
+	SystemDisk              *AlicloudSystemDisk
+	InstanceChargeType      string
+	InternetChargeType      string
+	InternetMaxBandwidthIn  *int
+	InternetMaxBandwidthOut *int
+	SpotStrategy            string
+	IoOptimized             string
+	Tags                    *AlicloudTags
+	KeyPairName             string
+	SecretRef               *corev1.SecretReference
+}
+
+// AlicloudSystemDisk describes SystemDisk for Alicloud.
+type AlicloudSystemDisk struct {
+	Category string
+	Size     int
+}
+
+// AlicloudTags describes Tags for Alicloud.
+type AlicloudTags struct {
+	Tag1Key   string
```

Comment:
Would it make more sense to have a list of tags rather than having one for each variable?

Reply:
We use this format because Alicloud doesn't support nested tags in its API (refer to https://help.aliyun.com/document_detail/63440.html, and search key word "Tag.n.Key"). We can make it as a list, but need extra logic of converting formats and validating list length.

Reply:
Hi @lu-zhenyu ,
The reason I am asking you to switch to list is bcoz we are trying to align all our machine classes into a single one, and when that happens we will need a list. Hence maybe you can read as a list and convert it while passing it to the SDK?

Refer - https://github.com/gardener/machine-controller-manager/blob/master/pkg/apis/machine/v1alpha1/types.go#L773, https://github.com/gardener/machine-controller-manager/blob/master/kubernetes/aws-machine-class.yaml#L19-L23

Reply:
noted, will do.